secrets/secrets.go

```diff
@@ -0,0 +1,172 @@
+package secrets
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"sync"
+)
+
+var (
+	ErrSecretsNotFound = errors.New("Secrets Implementor not registerd")
+	secretKeyOkMsg     = "Secret Key set successfully"
+	secretLoginOkMsg   = "Login completd"
+)
+
+type SecretManager interface {
+	// Login create session with secret store
+	SecretLogin(secretType int, secretConfig map[string]string) error
+	// SetClusterSecretKey sets the cluster wide secret key
+	SetClusterSecretKey(secretKey string, override bool) error
+	//CheckSecretLogin validates session with secret store
+	CheckSecretLogin() error
+	//SetSecret sets secret key against data
+	SetSecret(key string, value string) error
+	// GetSecret retrieves the data for the given feature and key
+	GetSecret(key string) (string, error)
+	//Routes to the secret API
+	Routes() []*Route
+}
+
+type Manager struct {
+	secret SecretManager
+	lock   sync.Mutex
+}
+
+func NewSecretManager(sec SecretManager) *Manager {
+	return &Manager{
+		secret: sec,
+	}
+}
+
+func (m *Manager) Routes() []*Route {
+	return []*Route{
+		{Verb: "GET", Path: secretVersion("checksecretslogin", APIVersion), Fn: m.secretLoginCheck},
+		{Verb: "GET", Path: secretVersion("getsecrets", APIVersion), Fn: m.getSecrets},
+		{Verb: "PUT", Path: secretVersion("setsecrets", APIVersion), Fn: m.setSecrets},
+		{Verb: "POST", Path: secretVersion("setclustersecretkey", APIVersion), Fn: m.setClusterSecretKey},
+		{Verb: "POST", Path: secretVersion("secretslogin", APIVersion), Fn: m.secretsLogin},
+	}
+}
+func secretVersion(route, version string) string {
+	if version == "" {
+		return "/" + route
+	} else {
+		return "/" + version + "/" + route
+	}
+}
+
+//TODO: Add swagger yaml
+func (m *Manager) setClusterSecretKey(w http.ResponseWriter, r *http.Request) {
```

Comment:
This is not the way it should be done. It should be:

```
[ route ] 
    --> [ server handler function (generic) which then calls the interface implementation ] 
          --> [ SecretManager (in this file, it would the fake one you have) ]
```

Reply:
To summarise : 
Routes:
```{verb: "GET", path: clusterVersion("checksecretslogin", cluster.APIVersion), fn: c.secretLoginCheck}```
https://github.com/ram-infrac/openstorage/commit/4078f23de524b46b22ef314df58bb20a1a6d05fe#diff-17b630a1004c86897f2252802622cf29R34.
 
Handler:
```func (c *clusterApi) setClusterSecretKey(w http.ResponseWriter, r *http.Request) {```

Interface implementation called from handler as,
```secretValue, err := SecretManager.Secret.GetSecret(secretid)```
https://github.com/ram-infrac/openstorage/commit/4078f23de524b46b22ef314df58bb20a1a6d05fe#diff-b6913eea9e933617095ce95032187c81R109

If anyone wishes to override his own secret implementer he can 
```clusterApi.RegisterSecretManager(fake.New())```
https://github.com/ram-infrac/openstorage/commit/4078f23de524b46b22ef314df58bb20a1a6d05fe#diff-44a527ea72ea9572e1cf71c7f1e2a0f0R121